Swdb: Better error handling

lukash opened the pull request:
Don't close SQLite3 database if it wasn't opened (results in a misleading error message).

For unprivileged user, don't try to create the database - the assumption is the user never writes to the database and shouldn't have the permissions either.

Commit:
SQLite3: Do not close the database if it wasn't opened (RhBug:1761976)

If opening the database failed, don't close it, this causes an API
misuse error and hides the original error that happened during opening.

https://bugzilla.redhat.com/show_bug.cgi?id=1761976

lukash commented:
The tests are here: https://github.com/rpm-software-management/ci-dnf-stack/pull/816

j-mracek commented:
@lukash I am really sorry, but it looks like that the issue is already fixed in dnf-4.2.21. Please could you check https://github.com/rpm-software-management/libdnf/pull/890?

lukash commented:
> @lukash I am really sorry, but it looks like that the issue is already fixed in dnf-4.2.21. Please could you check #890?

I am aware of that, this fixes a different issue. It's described in the commits, if anything is missing in the explanation, let me know.

j-mracek commented:
@lukash The commit `SQLite3: Do not close the database if it wasn't opened (RhBug:1761976)`  is ready to merge. With other one I have a problem.
I believe that bz#1761976 is already resolved in 0.46.2. 
The most critical part of the second commit is not in the code, but in description.
The most important reason why to not let un-privilage user to create database is that the database will not contain old data, but then the migration will never reappear. The other reasons like that check of permissions could fail is misleading. The statement `Creating a new database should not be possible for an unprivileged user anyway` has limited informative value.

lukash commented:
@j-mracek sorry, but I don't understand what you mean.

> The most critical part of the second commit is not in the code, but in description.

Well, the description can be fixed, so I guess you do have a problem with the code...

> The most important reason why to not let un-privilage user to create database is that the database will not contain old data, but then the migration will never reappear.

But I am actually removing the code branch that would allow an unprivileged user to create the database...

> The other reasons like that check of permissions could fail is misleading. The statement `Creating a new database should not be possible for an unprivileged user anyway` has limited informative value.

You should probably be more descriptive or suggest how to fix the description (if I understand correctly that's what you don't like here).

FWIW I'd think that the reasons for an unprivileged user not being able to modify system files are obvious.

j-mracek commented:
Ok, first remove reference for the bug. The new commit message could be ...
```
 swdb: Don't create a new history DB for unprivileged user

The database created by unprivileged did not contained any older
data, therefore there is no difference between new database or
database created in memory.
Additionally the creation of the database prevents any older data
conversion from older type of databases.
```

lukash commented:
@j-mracek I've added a paragraph mentioning the points you described to the commit description. I've kept the bug reference though. While it is not fixing that error exactly, without this commit the error message for that case would still be wrong and misleading. I can still remove the bug reference if you insist, I just consider it relevant...

j-mracek commented:
@lukash If bug will be issued for RHEL the reference for the bug will be not possible. In this case it is only incorrect.

From commit message please remove `The check itself may fail on
isufficient permissions, breaking the logic that was in place before
this commit.`, because it is misleading. 

Please also remove `Creating a new database should not be possible for an unprivileged user
anyway, and all operations that modify the system (and therefore would
write to the history) are checking for the user being root too.`. In future it will be required to install content into a user space including modification of DB.

Please could you also modify `Thew new empty database could potentially also
prevent conversion of the old database format if it is present.` to something like: `Thew new empty database would also
prevent conversion of the old databases format if they are present.` Not only history db is converted but also yumDB.

lukash commented:
> @lukash If bug will be issued for RHEL the reference for the bug will be not possible. In this case it is only incorrect.

I don't really understand, but I've removed the bug reference...

The following is just splitting hairs, but anyway...

> From commit message please remove `The check itself may fail on isufficient permissions, breaking the logic that was in place before this commit.`, because it is misleading.

In my opinion this line actually describes the problem itself. If you say it's misleading, you should also describe in which way, so that I can fix it. I put the description of the problem there on purpose.

> Please also remove `Creating a new database should not be possible for an unprivileged user anyway, and all operations that modify the system (and therefore would write to the history) are checking for the user being root too.`. In future it will be required to install content into a user space including modification of DB.

I am describing the current state of things as of this commit. What will happen in the future will be described in future commits, as of now, my statement holds true and explains why the change is correct under current circumstances.

> Please could you also modify `Thew new empty database could potentially also prevent conversion of the old database format if it is present.` to something like: `Thew new empty database would also prevent conversion of the old databases format if they are present.` Not only history db is converted but also yumDB.

I've changed the singular to plural...

j-mracek commented:
> The check itself may fail on isufficient permissions
This part tells me that something wrong is with check that the file is present. But there is no issue with the check.  Also word fail could be explain by multiple way. Previously you reference the bug reported a crash, therefore the fail could be also explain (incorrectly) as a crash. The most critical part was that you removed incorrect reference for the bug. With the rest I do not agree but I can merge it.

lukash commented:
> > The check itself may fail on isufficient permissions

> This part tells me that something wrong is with check that the file is present. But there is no issue with the check.  Also word fail could be explain by multiple way. ...

The issue is, if you have insufficient permissions to read the database file, the `pathExists(path.c_str())` check itself will fail on EPERM (I think that's the code you get). However, the function just ignores any errors and tells you the file doesn't exist. That breaks the logic, you think the file isn't there but it is, you just don't have permissions. That causes a wrong error to be reported to the user.

I guess I failed to mention that detail about `pathExists(path.c_str())`, I've added it to the commit description. Hope it's clear now?

j-mracek commented:
@rh-atomic-bot try

j-mracek commented:
@rh-atomic-bot retry

j-mracek commented:
@lukash Please could you check the output from RPM buil?

lukash commented:
Tests were broken in `dnf-4-master`.

@rh-atomic-bot retry

lukash commented:
@rh-atomic-bot retry

lukash commented:
@rh-atomic-bot retry

m-blaha commented:
Hm, dnf unit tests are failing. Let's look if it is another bug in CI...

m-blaha commented:
They are failing also locally so the failure is probably caused by this PR.

lukash commented:
> They are failing also locally so the failure is probably caused by this PR.

Yeah, I've had a look yesterday but it's a lot of weird failures, while CI tests work... I'll investigate it.

j-mracek commented:
I consider failing unittests as a problem.

lukash commented:
> I consider failing unittests as a problem.

Oh. really? :yum:

Commit:
utils/sqlite3: Throw SQLite3::Error if Statement creation failed

In case creating the statement failed, we can't throw Statement::Error,
the statement isn't there.

Commit:
Don't create a new history DB connection for in-memory DB

The hack that opens a new history DB connection can't work for an
in-memory database, as it will create a new, un-initialized database.

Commit:
utils: Add a safe version of pathExists()

Add pathExistsSafe(), which will throw an exception in case an error
occurred while accesing the path, other than ENOENT (i.e. file doesn't
exist).

Commit:
swdb: Handle the case when pathExists() fails on e.g. permission

When running dnf under an unprivileged user, make sure the check for the
existence of the history.sqlite database file is not failing on
something else, typically permissions. This is to report the proper
error message to the user. Previously, the user would get a message the
history.sqlite file doesn't exist even if it did and it was reading it
that failed.

Commit:
transaction/Swdb: Use a single logger variable in constructor

It's going to be optimized out by the compiler anyway, no need to
duplicate the line.

lukash commented:
@j-mracek I've had to put back the creation of the database for unprivileged user, the dnf unit tests just wouldn't work without it (I gave up finding out why). So the new version is much closer to the original and I've added a safe version of the `pathExists()` check that will make sure the correct error message is printed.

There is also a number of other fixes for issues I've encountered along the way...

And the tests are updated too.

j-mracek commented:
@lukash Shall I understand it that the real issue (creation of empty DB without data conversion) is not solved, but the pull request only improves an error output in very rare situation?

j-mracek commented on libdnf/transaction/Swdb.cpp:
```diff
@@ -71,7 +91,6 @@ Swdb::Swdb(const std::string &path)
                 conn->exec("BEGIN; UPDATE config SET value='test' WHERE key='test'; ROLLBACK;");
             } catch (SQLite3::Error & ex) {
                 // root must have the database writable -> log and re-throw the exception
-                auto logger(libdnf::Log::getLogger());
```
Why did you replace it? I believe that it will cause additional problem, because the logger will be loaded before it is fully initialized.

lukash replied:
It won't. The getter doesn't do any loading, it simply returns a static variable. By moving it to the beginning of the constructor doesn't really change anything. The getter would simply be called a bit earlier in the constructor, but I think it's going to be optimized out by the compiler anyway.

j-mracek replied:
The difference is really minor. It means that the logger is always imported even when it is not required. It means that the code will be slightly slower. The change has no benefit.

lukash replied:
What do you mean by "imported"? There is no difference at all. The getter only returns a pointer and that is stored in a local variable. There is absolutely no semantic difference and as I've said already, the getter itself is most likely optimized out by the compiler.

In case the getter wouldn't be optimized out, the performance impact is totally negligible. The benefit is saving multiple lines with the same content in the source.

Unless I'm missing something, it's really quite pointless arguing about this.

lukash commented:
> @lukash Shall I understand it that the real issue (creation of empty DB without data conversion) is not solved, but the pull request only improves an error output in very rare situation?

This is confusing. Creation of an empty DB (without the schema, I haven't been solving conversion issues at all) was an issue I've encountered while debugging the dnf unit tests and I've fixed that - AFAIK. The tests were still failing (when using in-memory DB) because they are closing and reopening the DB, which meant the data was disappearing - which doesn't happen if a file is used.

This PR has been about fixing an error message since the beginning, but it is not a very rare situation, it is the situation from the bug report.

j-mracek commented on libdnf/utils/filesystem.hpp:
```diff
@@ -30,6 +30,9 @@ namespace libdnf {
 pathExists(const char *path);
 
+bool
+pathExistsSafe(const std::string & path);
```
Please could you add a description here. If you have only devel package available, then the description is hidden. Probably using word Safe is not the perfect one. Please could you find a better name?

lukash replied:
I'll add the description. Regarding the name, I've already thought about it and this is the best I have come up with, suggestions welcome.

j-mracek replied:
pathExists2

j-mracek commented on libdnf/transaction/CompsEnvironmentItem.cpp:
```diff
@@ -187,8 +187,10 @@ CompsEnvironmentItem::getTransactionItemsByPattern(SQLite3Ptr conn, const std::s
     // after forking process in Anaconda
-    SQLite3 privateConn(conn->getPath());
-    SQLite3::Query query(privateConn, sql);
+    if (conn->getPath() != ":memory:") {
```
Is there any report that it can happen?

lukash replied:
Under a regular user that doesn't have the permissions to access the history database (either is not readable or doesn't exist and they don't have write permissions). Realistically it's a rare corner case but it can happen.

j-mracek commented on libdnf/utils/filesystem.cpp:
```diff
@@ -39,6 +42,24 @@ pathExists(const char *path)
+pathExistsSafe(const std::string & path)
+{
+    struct stat buffer;
+    if (stat(path.c_str(), &buffer) == 0) return true;
```
Please could you split it into two lines?

lukash commented:
@j-mracek I've added a block to the `if` body and I've moved the documentation of `pathExistsSave()` to the header.

j-mracek commented:
Tested locally.
